From b0705462f569e3f820ee30eb251880b651fe7b56 Mon Sep 17 00:00:00 2001
From: Sergey 'Jin' Bostandzhyan <jin@mediatomb.cc>
Date: Mon, 11 May 2020 17:38:41 +0200
Subject: [PATCH 1/3] Attempt to package klippy using setuptools

I am totally unfamiliar with python packaging, so not sure if there is a
better way, but at least it seems to work.

Custom compilation of c module code was removed, setuptools will take
care of that and also allow to cross compile the module.
---
 klippy/__init__.py                |  7 +++++++
 klippy/chelper/__init__.py        | 17 -----------------
 klippy/extras/bed_screws.py       |  2 +-
 klippy/extras/bed_tilt.py         |  3 ++-
 klippy/extras/bltouch.py          |  5 +++--
 klippy/extras/bus.py              |  2 +-
 klippy/extras/delta_calibrate.py  |  3 ++-
 klippy/extras/display/uc1701.py   |  3 ++-
 klippy/extras/endstop_phase.py    |  2 +-
 klippy/extras/extruder_stepper.py |  2 +-
 klippy/extras/force_move.py       |  2 +-
 klippy/extras/manual_probe.py     |  2 +-
 klippy/extras/manual_stepper.py   |  3 ++-
 klippy/extras/probe.py            |  3 ++-
 klippy/extras/replicape.py        |  3 ++-
 klippy/extras/sx1509.py           |  2 +-
 klippy/extras/z_tilt.py           |  3 ++-
 klippy/kinematics/cartesian.py    |  2 +-
 klippy/kinematics/corexy.py       |  2 +-
 klippy/kinematics/delta.py        |  2 +-
 klippy/kinematics/extruder.py     |  2 +-
 klippy/kinematics/polar.py        |  2 +-
 klippy/kinematics/rotary_delta.py |  2 +-
 klippy/kinematics/winch.py        |  2 +-
 klippy/klippy.py                  |  2 +-
 klippy/toolhead.py                |  2 +-
 setup.py                          | 27 +++++++++++++++++++++++++++
 27 files changed, 67 insertions(+), 42 deletions(-)
 create mode 100644 klippy/__init__.py
 create mode 100644 setup.py

diff --git a/klippy/__init__.py b/klippy/__init__.py
new file mode 100644
index 00000000..b3884662
--- /dev/null
+++ b/klippy/__init__.py
@@ -0,0 +1,7 @@
+# Package definition for the extras directory
+#
+# Copyright (C) 2018  Kevin O'Connor <kevin@koconnor.net>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+
+__all__ = [ 'chelper', 'extras', 'kinematics' ]
diff --git a/klippy/chelper/__init__.py b/klippy/chelper/__init__.py
index 1a4b2d74..14e115f8 100644
--- a/klippy/chelper/__init__.py
+++ b/klippy/chelper/__init__.py
@@ -162,20 +162,6 @@ def get_mtimes(srcdir, filelist):
         out.append(t)
     return out
 
-# Check if the code needs to be compiled
-def check_build_code(srcdir, target, sources, cmd, other_files=[]):
-    src_times = get_mtimes(srcdir, sources + other_files)
-    obj_times = get_mtimes(srcdir, [target])
-    if not obj_times or max(src_times) > min(obj_times):
-        logging.info("Building C code module %s", target)
-        srcfiles = [os.path.join(srcdir, fname) for fname in sources]
-        destlib = os.path.join(srcdir, target)
-        res = os.system(cmd % (destlib, ' '.join(srcfiles)))
-        if res:
-            msg = "Unable to build C code module (error=%s)" % (res,)
-            logging.error(msg)
-            raise Exception(msg)
-
 FFI_main = None
 FFI_lib = None
 pyhelper_logging_callback = None
@@ -185,8 +171,6 @@ def get_ffi():
     global FFI_main, FFI_lib, pyhelper_logging_callback
     if FFI_lib is None:
         srcdir = os.path.dirname(os.path.realpath(__file__))
-        check_build_code(srcdir, DEST_LIB, SOURCE_FILES, COMPILE_CMD
-                         , OTHER_FILES)
         FFI_main = cffi.FFI()
         for d in defs_all:
             FFI_main.cdef(d)
@@ -213,7 +197,6 @@ HC_CMD = "sudo %s/hub-ctrl -h 0 -P 2 -p %d"
 def run_hub_ctrl(enable_power):
     srcdir = os.path.dirname(os.path.realpath(__file__))
     hubdir = os.path.join(srcdir, HC_SOURCE_DIR)
-    check_build_code(hubdir, HC_TARGET, HC_SOURCE_FILES, HC_COMPILE_CMD)
     os.system(HC_CMD % (hubdir, enable_power))
 
 
diff --git a/klippy/extras/bed_screws.py b/klippy/extras/bed_screws.py
index c10a791f..ee5a13ca 100644
--- a/klippy/extras/bed_screws.py
+++ b/klippy/extras/bed_screws.py
@@ -3,7 +3,7 @@
 # Copyright (C) 2019  Kevin O'Connor <kevin@koconnor.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
-import homing
+from klippy import homing
 
 def parse_coord(config, param):
     pair = config.get(param).strip().split(',', 1)
diff --git a/klippy/extras/bed_tilt.py b/klippy/extras/bed_tilt.py
index 910b40a9..7b66f759 100644
--- a/klippy/extras/bed_tilt.py
+++ b/klippy/extras/bed_tilt.py
@@ -4,7 +4,8 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging
-import probe, mathutil
+import probe
+from klippy import mathutil
 
 class BedTilt:
     def __init__(self, config):
diff --git a/klippy/extras/bltouch.py b/klippy/extras/bltouch.py
index 8e9166bd..1c2bb094 100644
--- a/klippy/extras/bltouch.py
+++ b/klippy/extras/bltouch.py
@@ -3,8 +3,9 @@
 # Copyright (C) 2018-2020  Kevin O'Connor <kevin@koconnor.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
-import logging
-import homing, probe
+import math, logging
+from klippy import homing
+import probe
 
 SIGNAL_PERIOD = 0.020
 MIN_CMD_TIME = 5 * SIGNAL_PERIOD
diff --git a/klippy/extras/bus.py b/klippy/extras/bus.py
index 7c6ae440..91c84a4e 100644
--- a/klippy/extras/bus.py
+++ b/klippy/extras/bus.py
@@ -3,7 +3,7 @@
 # Copyright (C) 2018,2019  Kevin O'Connor <kevin@koconnor.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
-import mcu
+from klippy import mcu
 
 def resolve_bus_name(mcu, param, bus):
     # Find enumerations for the given bus
diff --git a/klippy/extras/delta_calibrate.py b/klippy/extras/delta_calibrate.py
index abea7cc6..2e6c862a 100644
--- a/klippy/extras/delta_calibrate.py
+++ b/klippy/extras/delta_calibrate.py
@@ -4,7 +4,8 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import math, logging, collections
-import probe, mathutil
+import probe
+from klippy import mathutil
 
 # A "stable position" is a 3-tuple containing the number of steps
 # taken since hitting the endstop on each delta tower.  Delta
diff --git a/klippy/extras/display/uc1701.py b/klippy/extras/display/uc1701.py
index f19d9f97..b5b1cf10 100644
--- a/klippy/extras/display/uc1701.py
+++ b/klippy/extras/display/uc1701.py
@@ -5,7 +5,8 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging
-import icons, font8x14, extras.bus
+import icons, font8x14
+from klippy.extras import bus
 
 BACKGROUND_PRIORITY_CLOCK = 0x7fffffff00000000
 
diff --git a/klippy/extras/endstop_phase.py b/klippy/extras/endstop_phase.py
index 8cfea7fd..f6e75fe2 100644
--- a/klippy/extras/endstop_phase.py
+++ b/klippy/extras/endstop_phase.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import math, logging
-import homing
+from klippy import homing
 
 TRINAMIC_DRIVERS = ["tmc2130", "tmc2208", "tmc2209", "tmc2660", "tmc5160"]
 
diff --git a/klippy/extras/extruder_stepper.py b/klippy/extras/extruder_stepper.py
index f6d0e9ea..74b2e956 100644
--- a/klippy/extras/extruder_stepper.py
+++ b/klippy/extras/extruder_stepper.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging
-import stepper
+from klippy import stepper
 
 class ExtruderStepper:
     def __init__(self, config):
diff --git a/klippy/extras/force_move.py b/klippy/extras/force_move.py
index 647b0b33..b2247833 100644
--- a/klippy/extras/force_move.py
+++ b/klippy/extras/force_move.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import math, logging
-import chelper
+from klippy import chelper
 
 BUZZ_DISTANCE = 1.
 BUZZ_VELOCITY = BUZZ_DISTANCE / .250
diff --git a/klippy/extras/manual_probe.py b/klippy/extras/manual_probe.py
index 663f3c8e..f696a0ae 100644
--- a/klippy/extras/manual_probe.py
+++ b/klippy/extras/manual_probe.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging, bisect
-import homing
+from klippy import homing
 
 class ManualProbe:
     def __init__(self, config):
diff --git a/klippy/extras/manual_stepper.py b/klippy/extras/manual_stepper.py
index c549a951..92ac24a3 100644
--- a/klippy/extras/manual_stepper.py
+++ b/klippy/extras/manual_stepper.py
@@ -3,7 +3,8 @@
 # Copyright (C) 2019  Kevin O'Connor <kevin@koconnor.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
-import stepper, homing, force_move, chelper
+import force_move
+from klippy import stepper, homing, chelper
 
 ENDSTOP_SAMPLE_TIME = .000015
 ENDSTOP_SAMPLE_COUNT = 4
diff --git a/klippy/extras/probe.py b/klippy/extras/probe.py
index e58b6080..121efe78 100644
--- a/klippy/extras/probe.py
+++ b/klippy/extras/probe.py
@@ -4,7 +4,8 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging
-import pins, homing, manual_probe
+from klippy import pins, homing
+import manual_probe
 
 HINT_TIMEOUT = """
 Make sure to home the printer before probing. If the probe
diff --git a/klippy/extras/replicape.py b/klippy/extras/replicape.py
index 5938502e..907cb46f 100644
--- a/klippy/extras/replicape.py
+++ b/klippy/extras/replicape.py
@@ -4,7 +4,8 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging, os
-import pins, mcu, bus
+from klippy import pins, mcu
+import bus
 
 REPLICAPE_MAX_CURRENT = 3.84
 REPLICAPE_PCA9685_BUS = 2
diff --git a/klippy/extras/sx1509.py b/klippy/extras/sx1509.py
index 27f335c9..3f38337c 100644
--- a/klippy/extras/sx1509.py
+++ b/klippy/extras/sx1509.py
@@ -3,7 +3,7 @@
 # Copyright (C) 2018  Florian Heilmann <Florian.Heilmann@gmx.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
-import pins
+from klippy import pins
 import bus
 
 # Word registers
diff --git a/klippy/extras/z_tilt.py b/klippy/extras/z_tilt.py
index ef69c0e6..00b19699 100644
--- a/klippy/extras/z_tilt.py
+++ b/klippy/extras/z_tilt.py
@@ -4,7 +4,8 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging
-import probe, mathutil
+import probe
+from klippy import mathutil
 
 class ZAdjustHelper:
     def __init__(self, config, z_count):
diff --git a/klippy/kinematics/cartesian.py b/klippy/kinematics/cartesian.py
index df57edce..b400bd1d 100644
--- a/klippy/kinematics/cartesian.py
+++ b/klippy/kinematics/cartesian.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging
-import stepper, homing
+from klippy import stepper, homing
 
 class CartKinematics:
     def __init__(self, toolhead, config):
diff --git a/klippy/kinematics/corexy.py b/klippy/kinematics/corexy.py
index f04b3009..2da29c97 100644
--- a/klippy/kinematics/corexy.py
+++ b/klippy/kinematics/corexy.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging, math
-import stepper, homing
+from klippy import stepper, homing
 
 class CoreXYKinematics:
     def __init__(self, toolhead, config):
diff --git a/klippy/kinematics/delta.py b/klippy/kinematics/delta.py
index 885ce940..1c334547 100644
--- a/klippy/kinematics/delta.py
+++ b/klippy/kinematics/delta.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import math, logging
-import stepper, homing, mathutil
+from klippy import stepper, homing, mathutil
 
 # Slow moves once the ratio of tower to XY movement exceeds SLOW_RATIO
 SLOW_RATIO = 3.
diff --git a/klippy/kinematics/extruder.py b/klippy/kinematics/extruder.py
index c0ba299c..d314cdc3 100644
--- a/klippy/kinematics/extruder.py
+++ b/klippy/kinematics/extruder.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import math, logging
-import stepper, homing, chelper
+from klippy import stepper, homing, chelper
 
 class PrinterExtruder:
     def __init__(self, config, extruder_num):
diff --git a/klippy/kinematics/polar.py b/klippy/kinematics/polar.py
index e3ee3241..f2643552 100644
--- a/klippy/kinematics/polar.py
+++ b/klippy/kinematics/polar.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import logging, math
-import stepper, homing
+from klippy import stepper, homing
 
 class PolarKinematics:
     def __init__(self, toolhead, config):
diff --git a/klippy/kinematics/rotary_delta.py b/klippy/kinematics/rotary_delta.py
index 415a2e7f..ff62ba49 100644
--- a/klippy/kinematics/rotary_delta.py
+++ b/klippy/kinematics/rotary_delta.py
@@ -4,7 +4,7 @@
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
 import math, logging
-import stepper, homing, mathutil, chelper
+from klippy import stepper, homing, mathutil, chelper
 
 class RotaryDeltaKinematics:
     def __init__(self, toolhead, config):
diff --git a/klippy/kinematics/winch.py b/klippy/kinematics/winch.py
index 04e2d498..4de65d58 100644
--- a/klippy/kinematics/winch.py
+++ b/klippy/kinematics/winch.py
@@ -3,7 +3,7 @@
 # Copyright (C) 2018-2019  Kevin O'Connor <kevin@koconnor.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
-import stepper, mathutil
+from klippy import stepper, mathutil
 
 class WinchKinematics:
     def __init__(self, toolhead, config):
diff --git a/klippy/klippy.py b/klippy/klippy.py
index 84450aea..091dbc3c 100644
--- a/klippy/klippy.py
+++ b/klippy/klippy.py
@@ -106,7 +106,7 @@ class Printer:
             if default is not configfile.sentinel:
                 return default
             raise self.config_error("Unable to load module '%s'" % (section,))
-        mod = importlib.import_module('extras.' + module_name)
+        mod = importlib.import_module('klippy.extras.' + module_name)
         init_func = 'load_config'
         if len(module_parts) > 1:
             init_func = 'load_config_prefix'
diff --git a/klippy/toolhead.py b/klippy/toolhead.py
index cf8fba4a..e811a467 100644
--- a/klippy/toolhead.py
+++ b/klippy/toolhead.py
@@ -250,7 +250,7 @@ class ToolHead:
         self.extruder = kinematics.extruder.DummyExtruder()
         kin_name = config.get('kinematics')
         try:
-            mod = importlib.import_module('kinematics.' + kin_name)
+            mod = importlib.import_module('klippy.kinematics.' + kin_name)
             self.kin = mod.load_kinematics(self, config)
         except config.error as e:
             raise
diff --git a/setup.py b/setup.py
new file mode 100644
index 00000000..faa2d895
--- /dev/null
+++ b/setup.py
@@ -0,0 +1,27 @@
+from setuptools import setup, find_packages, Extension
+
+c_helper = Extension('klippy/chelper/c_helper',
+                    define_macros = [('MAJOR_VERSION', '1'),
+                                     ('MINOR_VERSION', '0')],
+                    sources = [ 'klippy/chelper/itersolve.c',
+                                'klippy/chelper/kin_cartesian.c',
+                                'klippy/chelper/kin_corexy.c',
+                                'klippy/chelper/kin_delta.c',
+                                'klippy/chelper/kin_extruder.c',
+                                'klippy/chelper/kin_polar.c',
+                                'klippy/chelper/kin_rotary_delta.c',
+                                'klippy/chelper/kin_winch.c',
+                                'klippy/chelper/pyhelper.c',
+                                'klippy/chelper/serialqueue.c',
+                                'klippy/chelper/stepcompress.c',
+                                'klippy/chelper/trapq.c' ])
+
+setup(
+    name = 'klipper',
+    version = '0.8.0+git',
+    description = 'Klipper 3D Printing Firmware',
+    packages=find_packages(),
+    entry_points = { "console_scripts": [ "klippy = klippy.klippy:main" ] },
+    url = "https://www.klipper3d.org/",
+    ext_modules = [ c_helper ]
+)
-- 
2.25.4

